### PR TITLE
rpc: have raw transaction decoding infer output descriptors

### PR DESCRIPTION
**Pull request by instagibbs**
Following discussion in #16725 this is complementary data to expose. All outputs are inferred.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
that's cool utACK

**Comment by meshcollider** [comment]
utACK c772728ab4a40bfb9a546eef669b0d7603bd9f89

**Comment by instagibbs** [comment]
rebased

**Comment by meshcollider** [comment]
You need to update the help-text for the RPC affected by this change (decoderawtransaction) to include the additional field

**Comment by laanwj** [comment]
Travis found a real problem here (for a change !!! :tada: )
```
/usr/bin/python3.6 ../test/util/bitcoin-util-test.py

2019-09-10 01:34:27,547 - ERROR - Output data mismatch for txcreateoutpubkey1.json (format json)

2019-09-10 01:34:27,548 - ERROR - Output formatting mismatch for txcreateoutpubkey1.json:

*** txcreateoutpubkey1.json

--- returned

***************

*** 15,22 ****

              "scriptPubKey": {

                  "asm": "02a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397 OP_CHECKSIG",

                  "hex": "2102a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397ac",

!                 "type": "pubkey",

!                 "desc": "pk(02a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397)#rk5v7uqw"

              }

          }

      ],

--- 15,21 ----

              "scriptPubKey": {

                  "asm": "02a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397 OP_CHECKSIG",

                  "hex": "2102a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397ac",

!                 "type": "pubkey"

              }

          }

      ],

2019-09-10 01:34:30,728 - ERROR - FAILED_TESTCASES:

['Creates a new transaction with a single pay-to-pubkey output (output as json)']
```

**Comment by instagibbs** [comment]
fixed rebase error, now *all* outputs are inferred. added description to `decodescript`

File: src/core_write.cpp [hunk 1]
```diff
@@ -162,6 +163,7 @@ void ScriptPubKeyToUniv(const CScript& scriptPubKey,
 
     if (!ExtractDestinations(scriptPubKey, type, addresses, nRequired) || type == TX_PUBKEY) {
         out.pushKV("type", GetTxnOutputType(type));
+        out.pushKV("desc", InferDescriptor(scriptPubKey, DUMMY_SIGNING_PROVIDER)->ToString());
```
Comment (promag) [inline src/core_write.cpp]: `desc` is always present, could just at the beginning? Like before `asm`.

File: src/core_write.cpp [hunk 2]
```diff
@@ -173,6 +175,8 @@ void ScriptPubKeyToUniv(const CScript& scriptPubKey,
         a.push_back(EncodeDestination(addr));
     }
     out.pushKV("addresses", a);
+
```
Comment (promag) [inline src/core_write.cpp]: nit, remove empty line?

File: src/core_write.cpp [hunk 3]
```diff
@@ -9,6 +9,7 @@
 #include <key_io.h>
 #include <script/script.h>
 #include <script/standard.h>
+#include <script/descriptor.h>
```
Comment (promag) [inline src/core_write.cpp]: gentle nit, keep sorted.

File: src/core_write.cpp (hunk 1 above)
Comment (instagibbs) [reply src/core_write.cpp]: done

File: src/core_write.cpp (hunk 3 above)
Comment (instagibbs) [reply src/core_write.cpp]: done

File: src/core_write.cpp [hunk 4]
```diff
@@ -157,6 +158,8 @@ void ScriptPubKeyToUniv(const CScript& scriptPubKey,
     int nRequired;
 
     out.pushKV("asm", ScriptToAsmStr(scriptPubKey));
+    out.pushKV("desc", InferDescriptor(scriptPubKey, DUMMY_SIGNING_PROVIDER)->ToString());
```
Comment (ariard) [inline src/core_write.cpp]: Need to update help-debug of `gettxout` and also maybe `REST-interface.md` on `rest_getutxos`.

Do you want also to extend feature to `decodepsbt` ?

**Review by ariard — COMMENTED** [review]
ACK 9b94596 minus `gettouxt/rest_getuxtos` doc fixs

Tested getrawtransaction, decoderawtransaction, decodescript, work as expected.

File: src/core_write.cpp (hunk 4 above)
Comment (instagibbs) [reply src/core_write.cpp]: > Do you want also to extend feature to decodepsbt ?

going to take a little more thinking, so for now "no" :)

fixed up other missing documentation to best of knowledge

Comment (instagibbs) [reply src/core_write.cpp]: Oh I see `**ScriptToUniv**` is only used for PSBT decoding... so yeah I can do that real quick.

**Comment by instagibbs** [comment]
pushed update fixing @ariard nits

**Comment by theStack** [comment]
Being pedantic here, but wouldn't it be more logical to put the inferred output descriptor `"desc"` before `"asm"` within the `"scriptPubKey"` json object, to keep the abstraction level order strictly decreasing (address -> script -> raw bytes)? From a user perspective I'd find it nice to find the interpretation of a scriptPubKey right in the first line, and I only need to dig deeper into the next two lines if I want more low-level representation.

**Comment by sipa** [comment]
Concept ACK. We should do this for consistency, though in practice this won't add much (it'll always either be a `raw()` or `addr()` descriptor which doesn't convey anything, except for P2PK and bare multisig, which aren't used anymore).

**Comment by instagibbs** [comment]
rebased. 

@theStack would have re-arranged if I didn't have to regenerate the tests again :)

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -1038,6 +1038,7 @@ UniValue gettxout(const JSONRPCRequest& request)
                         {RPCResult::Type::OBJ, "scriptPubKey", "",
                             {
                                 {RPCResult::Type::STR_HEX, "asm", ""},
+                                {RPCResult::Type::STR_HEX, "desc", "Inferred descriptor for the output\n"},
```
Comment (luke-jr) [inline src/rpc/blockchain.cpp:1041]: Pretty sure these aren't hex?

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/rpc/blockchain.cpp (hunk 5 above)
Comment (instagibbs) [reply src/rpc/blockchain.cpp]: fixed

**Comment by instagibbs** [comment]
failure for createmultisig tests: https://travis-ci.org/github/bitcoin/bitcoin/jobs/716675920

presuming it's unrelated?

File: src/rpc/rawtransaction.cpp [hunk 6]
```diff
@@ -555,6 +557,7 @@ static UniValue decodescript(const JSONRPCRequest& request)
                             {
                                 {RPCResult::Type::STR, "address", "segwit address"},
                             }},
+                            {RPCResult::Type::STR_HEX, "desc", "Inferred descriptor for the output\n"},
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction.cpp:560]: why hex?

Comment (instagibbs) [reply src/rpc/rawtransaction.cpp:560]: it's not, forgot to change this when fixing the other one

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/rpc/rawtransaction.cpp [hunk 7]
```diff
@@ -555,6 +557,7 @@ static UniValue decodescript(const JSONRPCRequest& request)
                             {
                                 {RPCResult::Type::STR, "address", "segwit address"},
                             }},
+                            {RPCResult::Type::STR, "desc", "Inferred descriptor for the output\n"},
```
Comment (luke-jr) [inline src/rpc/rawtransaction.cpp:560]: There's no output here, just a raw script.

File: src/rpc/blockchain.cpp [hunk 8]
```diff
@@ -1038,6 +1038,7 @@ UniValue gettxout(const JSONRPCRequest& request)
                         {RPCResult::Type::OBJ, "scriptPubKey", "",
                             {
                                 {RPCResult::Type::STR_HEX, "asm", ""},
+                                {RPCResult::Type::STR, "desc", "Inferred descriptor for the output\n"},
```
Comment (luke-jr) [inline src/rpc/blockchain.cpp:1041]: There's no need for the `\n` in all these

**Comment by sipa** [comment]
It would be nice to have this, as it compensates somewhat for the (weird) functionality lost in #20286. There are a bunch of unaddressed comments here though. And also, it'd be good if the `decodescript` output inside the "p2sh" and "segwit" blocks would report a descriptor that takes the structure into account.

**Comment by instagibbs** [comment]
Ah, thought I closed this PR. I'll take a fresh look.

File: src/rpc/rawtransaction.cpp (hunk 7 above)
Comment (instagibbs) [reply src/rpc/rawtransaction.cpp:560]: can you give a suggestion?

Comment (sipa) [reply src/rpc/rawtransaction.cpp]: "Inferred output descriptor for the script" ?

Comment (instagibbs) [reply src/rpc/rawtransaction.cpp]: right I shortly realized what he meant, already pushed

**Comment by instagibbs** [comment]
@sipa Looks like it already does the native segwit inference in `decodescript`, `p2sh` doesn't have a block per se, just a single field with the p2sh address itself. Could add it I guess? Name suggestion for the object?

**Comment by sipa** [comment]
@instagibbs I was confusing `decodescript`'s "p2sh" and "segwit" fields with the "embedded" field in `getaddressinfo` (thinking it'd give an Object with fields for the p2sh/segwit version of the specified script, instead of just the resulting address). Perhaps there is a use for that, but not in this PR.

File: src/rpc/rawtransaction.cpp (hunk 7 above)
Comment (instagibbs) [reply src/rpc/rawtransaction.cpp]: fixed

**Comment by instagibbs** [comment]
Addressed all concerns, I think.

**Comment by FndNur1Labs** [comment]
Desc for string. Update is good. Need known rebase that.

**Comment by instagibbs** [comment]
kk will rebase

**Comment by instagibbs** [comment]
@meshcollider ready for rereview

File: src/core_write.cpp [hunk 9]
```diff
@@ -152,8 +153,10 @@ void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool include
     CTxDestination address;
 
     out.pushKV("asm", ScriptToAsmStr(scriptPubKey));
+    out.pushKV("desc", InferDescriptor(scriptPubKey, DUMMY_SIGNING_PROVIDER)->ToString());
     if (include_hex) out.pushKV("hex", HexStr(scriptPubKey));
 
+
```
Comment (achow101) [inline src/core_write.cpp:159]: In c90a08886afe86d84bbd1d76ae3d3ecc24f24a74 "transaction decoding infer output descriptors"

Extraneous line

File: src/rpc/blockchain.cpp [hunk 10]
```diff
@@ -1331,6 +1331,7 @@ static RPCHelpMan gettxout()
                 {RPCResult::Type::STR_AMOUNT, "value", "The transaction value in " + CURRENCY_UNIT},
                 {RPCResult::Type::OBJ, "scriptPubKey", "", {
                     {RPCResult::Type::STR, "asm", ""},
+                                {RPCResult::Type::STR, "desc", "Inferred descriptor for the output\n"},
```
Comment (achow101) [inline src/rpc/blockchain.cpp:1334]: In c90a08886afe86d84bbd1d76ae3d3ecc24f24a74 "transaction decoding infer output descriptors"

Indentation, also an extraneous `\n`

File: src/rpc/rawtransaction.cpp [hunk 11]
```diff
@@ -561,6 +563,7 @@ static RPCHelpMan decodescript()
             RPCResult::Type::OBJ, "", "",
             {
                 {RPCResult::Type::STR, "asm", "Script public key"},
+                        {RPCResult::Type::STR, "desc", "Inferred descriptor for the script"},
```
Comment (achow101) [inline src/rpc/rawtransaction.cpp:566]: In c90a08886afe86d84bbd1d76ae3d3ecc24f24a74 "transaction decoding infer output descriptors"

Indentation

File: src/rpc/rawtransaction.cpp [hunk 12]
```diff
@@ -572,6 +575,7 @@ static RPCHelpMan decodescript()
                      {RPCResult::Type::STR_HEX, "hex", "Hex string of the script public key"},
                      {RPCResult::Type::STR, "type", "The type of the script public key (e.g. witness_v0_keyhash or witness_v0_scripthash)"},
                      {RPCResult::Type::STR, "address", /*optional=*/true, "The Bitcoin address (only if a well-defined address exists)"},
+                            {RPCResult::Type::STR, "desc", "Inferred descriptor for the script"},
```
Comment (achow101) [inline src/rpc/rawtransaction.cpp:578]: Indentation

**Comment by instagibbs** [comment]
rebased and formatting issues fixed

**Comment by achow101** [comment]
ACK 6498ba151b35ce9621ad00730f1fdfca55538ace

**Review by meshcollider — COMMENTED** [review]
utACK 6498ba151b35ce9621ad00730f1fdfca55538ace

**Comment by MarcoFalke** [comment]
I don't think we should be using output descriptors to describe redeem scripts and witness scripts. If you agree, this should be reverted or fixed before 23.0 final. If you disagree, pls leave a comment on https://github.com/bitcoin/bitcoin/pull/24636

File: src/core_write.cpp (hunk 4 above)
Comment (MarcoFalke) [reply src/core_write.cpp:156]: For reference, `decodepsbt` *is* included in this pull (despite the docs not being updated)

**Pull request merged**